scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,466 @@
+       S. Tarantola. "Variance based sensitivity analysis of model
+       output. Design and estimator for the total sensitivity index."
+       Computer Physics Communications, 181(2):259-270,
+       :doi:`10.1016/j.cpc.2009.09.018`, 2010.
+    """
+    f_AB = f_AB.reshape(-1, *f_A.shape)
+
+    var = np.var(np.vstack([f_A, f_B]), axis=0)
+
+    s = np.mean(f_B * (f_AB - f_A), axis=1) / var
+    st = 0.5 * np.mean((f_A - f_AB) ** 2, axis=1) / var
+
+    return s.T, st.T
+
+
+@dataclass
+class SobolResult:
+    first_order: np.ndarray
+    total_order: np.ndarray
+    _A: np.ndarray
+    _B: np.ndarray
+    _AB: np.ndarray
+    _f_A: np.ndarray
+    _f_B: np.ndarray
+    _f_AB: np.ndarray
+    _bootstrap_result: BootstrapResult = None  # type: ignore[valid-type]
+
+    def bootstrap(
+        self,
+        confidence_level: DecimalNumber = 0.95,
+        n_resamples: IntNumber = 99
+    ) -> BootstrapResult:  # type: ignore[valid-type]
+        """Bootstrap Sobol' indices to provide confidence intervals.
+
+        Parameters
+        ----------
+        confidence_level : float, default: ``0.95``
+            The confidence level of the confidence intervals.
+        n_resamples : int, default: ``99``
+            The number of resamples performed to form the bootstrap
+            distribution of the indices.
+
+        Returns
+        -------
+        res : BootstrapResult
+            Bootstrap result containing the confidence intervals and the
+            bootstrap distribution of the indices.
+
+        """
+        def statistic(idx):
+            f_A_ = self._f_A[idx]
+            f_B_ = self._f_B[idx]
+
+            n, d = self._A.shape
+            f_AB_ = np.empty((d, *f_A_.shape))
+            for i in range(d):
+                f_AB_[i] = self._f_AB[np.array(idx)+i*n]
+            f_AB_ = f_AB_.reshape(-1, f_A_.shape[1])
+
+            return saltelli_2010(f_A_, f_B_, f_AB_)
+
+        n = len(self._f_A)
+
+        self._bootstrap_result = bootstrap(
+            [np.arange(n)], statistic=statistic, method="BCa",
+            n_resamples=n_resamples,
+            confidence_level=confidence_level,
+            bootstrap_result=self._bootstrap_result
+        )
+
+        return self._bootstrap_result
+
+
+def sobol_indices(
+    *,
+    func: Callable,
+    n: IntNumber,
+    dists: List[PINVDist],
+    random_state: SeedType = None
+) -> SobolResult:
+    r"""Global sensitivity indices of Sobol'.
+
+    Parameters
+    ----------
+    func : callable
+        Function to compute the Sobol' indices from. It's signature must be
+        ``func(x: ArrayLike) -> ArrayLike``, with ``x`` of shape ``(-1, d)``
+        and the output should have a shape ``(-1, s)`` with ``s`` the number
+        of output.
+    n : int
+        Must be a power of 2. The total number of function call will be
+        ``n(d+2)``.
+    dists : list(distributions), optional
+        List of distributions of the parameters. It must be compatible with
```

Comment:
I'm not sure that I'd know what "distributions of the parameters" would be in this context without reading the implementation code. The parameters don't "have distributions" per se. Rather, we're using this machinery to map the domain of the function onto the unit hypercube.

I'd rather go back to the list of bounds formulation. It's going to be by far the most common case; using `uniform(loc=lower_bound, scale=upper_bound - lower_bound)` is an unnecessarily circuitous way to specify that case. You can provide examples that wrap infinite-bounded functions using, say, `stats.norm.ppf()` to map the unit-hypercube inputs to the infinite bounds.

Or you can allow the list of bounds where each parameter can have either `(low, high)` bounds or an `rv_continuous` distribution object, and you can continue to take care of the mapping internally using `NumericalInversePolynomial`. But the way it's being used should be documented.

Reply:
Ok I will try to rephrase that. What I wanted to say was that each parameter from a user perspective is following a certain distribution. The unit hypercube is as you said just some internal machinery and a user should not even know about that.

On the bounds, I had a first design of the API with bounds and distributions but in the end I removed the bounds. To me it was duplicating the information as the distribution should already contain the bounds. And otherwise what does it mean if I have for e.g. a normal distribution and put some bounds? How do we apply the bounds? I didn't know how to mix this information properly.

In most well done analysis, people actually manage to have real meaningful distributions for their parameters. Putting a uniform distribution for a parameter is really telling, well here I have no clue so using this. The issue is that the choice of distribution has a huge impact on the analysis.

All in all, I really think it's a cleaner API and more close to how people use this method in real life. When talking to other practitioners, we always talk about the distributions we have for our parameters so it is very natural.